brainiak/matnorm_models/regression.py

```diff
@@ -0,0 +1,167 @@
+        """ Log likelihood of model (internal)
+        """
+        y_hat = tf.matmul(self.X, self.beta)
+        resid = self.Y - y_hat
+        return self.matnorm_logp(resid, self.time_noise_cov, self.space_noise_cov)
+
+    def fit(self, X, y, voxel_pos=None, times=None, max_iter=1000, step=100,
+            loss_tol=1e-8, grad_tol=1e-8, restart=True):
+        """ Compute the regression fit.
+
+        Parameters
+        ----------
+        X : np.array, TRs by conditions.
+            Design matrix
+        Y : np.array, TRs by voxels.
+            fMRI data
+        voxel_pos: np.array, n_voxels by 3, default: None
+            Spatial positions of voxels (optional).
+            If provided, and if space_noise_cov is a NoiseCovGP, the positions
+            for computing the GP covaraince matrix. Otherwise NoiseCovGP
+            defaults to distances of 1 unit between all voxels.
+            Ignored by non-GP noise covariances.
+        times : np.array, TRs by 1, default:None
+            Timestamps of observations (optional).
+            If provided, and if time_noise_cov is a NoiseCovGP, the the times
+            for computing the GP covaraince matrix. Otherwise NoiseCovGP
+            defaults to distances of 1 unit between all times.
+            Ignored by non-GP noise covariances.
+        max_iter: int, default=1000
+            Maximum number of iterations to run
+        step: int, default=100
+            Number of steps between optimizer status outputs.
+        restart: bool, default=True
+            If this is true, optimizer is restarted (e.g. for a new dataset).
+            Otherwise optimizer will continue from where it is now (for example
+            for running more iterations if the initial number was not enough).
+        """
+
+        if restart:
+            self.sess.run(tf.global_variables_initializer())
+
+        if voxel_pos is None:
+            self.voxel_pos = np.c_[np.arange(self.n_v)[:, None],
+                                   np.zeros((self.n_v, 2))]
+
+        if times is None:
+            times = np.arange(X.shape[0])[:, None]
+
+        feed_dict = {self.X: X, self.Y: y, self.time_noise_cov.loc: times,
+                     self.space_noise_cov.loc: self.voxel_pos}
+
+        self._optimize_impl(self.train_optimize, -self.train_logp,
+                            self.train_variables, feed_dict, max_iter,
+                            step, loss_tol, grad_tol)
+
+        self.beta_ = self.beta.eval(session=self.sess)
+
+    def predict(self, X):
+        """ Predict fMRI signal from design matrix.
+
+        Parameters
+        ----------
+        X : np.array, TRs by conditions.
+            Design matrix
+
+        """
+
+        y_predicted = tf.matmul(self.X, self.beta)
+
+        self.y_predicted_ = y_predicted.eval(session=self.sess, 
+                                             feed_dict={self.X: X})
+
+        return self.y_predicted_
+
+    def calibrate(self, Y):
```

Comment:
https://en.wikipedia.org/wiki/Calibration_(statistics)#In_regression is a term for finding X from Y after having fit a regression model Y ~ XB. Aka "inverse regression" or what fMRI people would call a decoder, but I didn't want to call it `decode` because this is also what fMRI people would call an encoding model. I'm open to other suggestions.